monkestation/code/modules/storytellers/event_groups/_event_group.dm

```diff
@@ -0,0 +1,60 @@
+/// An associative list of singleton event groups, in the format of [type] = instance.
+GLOBAL_LIST_INIT_TYPED(event_groups, /datum/event_group, initialize_event_groups())
+
+/datum/event_group
+	/// The name of the event group.
+	var/name
+	/// If set, this will limit the amount of times events in this group can run.
+	var/max_occurrences
+	/// If set, whenever an event in this group runs, no other events in this group
+	/// will be able to run for the specified amount of time.
+	/// Can be either a number for a static cooldown, or a list containing 2 numbers,
+	/// for a random cooldown between the given numbers, which will bias towards the upper bound
+	/// the more the event group has occured.
+	var/cooldown_time
+	/// The amount of time events in this group have ran.
+	VAR_FINAL/occurrences = 0
+	/// If [cooldown_time] is set, this will be set to the minimum world.time where the next event in this group can run.
+	COOLDOWN_DECLARE(event_cooldown)
+
+/datum/event_group/proc/get_cooldown_time() as num
+	if(isnum(cooldown_time))
+		return cooldown_time
+	var/min_cooldown = cooldown_time[1]
+	var/max_cooldown = cooldown_time[2]
+
+	if (max_occurrences)
+		var/occurrence_ratio = min(1, occurrences / max_occurrences)
+		return min_cooldown + (max_cooldown - min_cooldown) * occurrence_ratio
+	else
+		// If max_occurrences is not set, use a simple exponential increase
+		return min_cooldown + (max_cooldown - min_cooldown) * (occurrences / (occurrences + 1))
+
+/datum/event_group/proc/can_run() as num
```

Comment:
unsure the as num is needed here, although im not sure how much of an effect removing it would have

Reply:
it's completely unneeded, i've just made a habit of adding byond return types where possible